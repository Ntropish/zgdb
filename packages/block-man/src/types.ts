export type Address = Uint8Array;

export interface BlockStore {
  /**
   * Retrieves a block of data from the store.
   * @param address The content address of the block.
   * @returns A Promise that resolves to the block's data as a Uint8Array, or undefined if not found.
   */
  get: (address: Address) => Promise<Uint8Array | undefined>;

  /**
   * Stores a block of data.
   * @param block The block data to store.
   * @returns A Promise that resolves to the content address of the stored block.
   */
  put: (block: Uint8Array) => Promise<Address>;

  /**
   * Checks for the existence of a block.
   * @param address The content address of the block.
   * @returns A Promise that resolves to true if the block exists, false otherwise.
   */
  has: (address: Address) => Promise<boolean>;
}

export interface Hasher {
  (data: Uint8Array): Promise<Address>;
}
