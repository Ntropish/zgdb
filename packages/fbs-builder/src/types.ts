export type FbsScalar =
  | "bool"
  | "byte"
  | "ubyte"
  | "short"
  | "ushort"
  | "int"
  | "uint"
  | "long"
  | "ulong"
  | "float"
  | "double";

export type FbsType = FbsScalar | string; // string for other tables, structs, etc.

export interface FbsField {
  name: string;
  type: FbsType;
  defaultValue?: number | string | boolean | null;
  isVector: boolean;
  attributes: Map<string, string | boolean>;
}

export interface FbsDeclaration {
  kind: "table" | "struct" | "enum" | "union";
  name: string;
  docs: string[];
}

export interface FbsTableState extends FbsDeclaration {
  kind: "table";
  fields: FbsField[];
  attributes: Map<string, string | boolean>;
}

export interface FbsStructState extends FbsDeclaration {
  kind: "struct";
  fields: FbsField[]; // in structs, fields cannot be optional or have defaults
}

export interface FbsEnumState extends FbsDeclaration {
  kind: "enum";
  type: FbsScalar;
  values: Map<string, number>;
}

export interface FbsUnionState extends FbsDeclaration {
  kind: "union";
  values: string[]; // table names
}

export interface FbsFileState {
  namespace: string | null;
  includes: string[];
  declarations: FbsDeclaration[];
  rootType: string | null;
  fileIdentifier: string | null;
  fileExtension: string | null;
}
