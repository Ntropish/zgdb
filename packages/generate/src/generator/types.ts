import { NormalizedSchema } from "../parser/types.js";

export interface GeneratorConfig {
  schemas: NormalizedSchema[];
  outputDirectory: string;
  options: {
    /**
     * The file extension to use for the generated import of the FlatBuffers schema.
     * E.g., '.js', '.mjs', or an empty string.
     * @default ".js"
     */
    importExtension?: string;
  };
}
