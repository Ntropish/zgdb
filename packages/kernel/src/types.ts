import { ComponentFactory } from "./component";

/**
 * A unique symbol used to signal that an orchestrator pipeline should halt execution.
 */
export const BREAK = Symbol.for("Orchestrator.BREAK");

/**
 * The most basic execution unit in the kernel. A function that accepts a context object
 * and can perform synchronous or asynchronous work.
 */
export type StepHandler<TContext extends Record<string, any> = any> = (
  ctx: TContext
) => unknown;

/**
 * The interface for a kernel that executes a linear pipeline of StepHandlers.
 */
export interface OrchestratorKernel<
  TContext extends Record<string, any> = any
> {
  run(ctx?: Partial<TContext>): Promise<TContext | typeof BREAK>;
  clone(): OrchestratorKernel<TContext>;
}

/**
 * A reactor's event handler is a self-contained orchestrator pipeline.
 */
export type EventHandler<TContext extends Record<string, any> = any> =
  OrchestratorKernel<TContext>;

/**
 * A structured collection of StepHandler arrays for logging at different levels.
 * This is designed to be injected into kernels to provide logging capabilities
 * without creating circular dependencies.
 */
export type LoggerPlugins = {
  info?: StepHandler[];
  warn?: StepHandler[];
  error?: StepHandler[];
};
