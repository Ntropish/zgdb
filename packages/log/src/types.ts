export type LogLevel = "fatal" | "error" | "warn" | "info" | "debug" | "trace";

export const LogLevels: Record<LogLevel, number> = {
  fatal: 60,
  error: 50,
  warn: 40,
  info: 30,
  debug: 20,
  trace: 10,
};

export interface LogRecord {
  level: LogLevel;
  message: string;
  time: number;
  data?: Record<string, unknown>;
  error?: Error;
}

export interface Transport {
  send(payload: string): void | Promise<void>;
}
