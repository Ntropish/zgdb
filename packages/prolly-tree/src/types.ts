export type Diff = {
  key: Uint8Array;
  localValue?: Uint8Array;
  remoteValue?: Uint8Array;
};

export type ConflictResolver = (
  key: Uint8Array,
  base: Uint8Array | undefined,
  local: Uint8Array,
  remote: Uint8Array
) => Promise<Uint8Array>;
