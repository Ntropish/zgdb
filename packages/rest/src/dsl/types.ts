import { Orchestrator, StepHandler } from "@tsmk/kernel";
import { Schema } from "@tsmk/schema";
import { IncomingMessage, ServerResponse } from "http";

// This will be refined with a strongly-typed HttpContext.
export interface Context {
  request: IncomingMessage;
  response: ServerResponse & {
    body?: any;
    statusCode?: number;
    headers?: Record<string, string>;
  };
  params: Record<string, string>;
  [key: string]: any;
}

export type Next = () => Promise<void>;

export type Handler<T extends Context = Context> = (
  ctx: T,
  next: Next
) => Promise<void>;

export type ErrorHandler = (err: Error, ctx: any) => any;

export type { StepHandler };

export interface ServerDefinition {
  type: "server";
  middleware: StepHandler[];
  items: Array<GroupDefinition | RouteDefinition>;
}

export interface GroupDefinition {
  type: "group";
  path: string;
  middleware: StepHandler[];
  items: Array<GroupDefinition | RouteDefinition>;
}

export interface RouteDefinition {
  type: "route";
  method: "GET" | "POST" | "PUT" | "PATCH" | "DELETE";
  path: string;
  handler: StepHandler;
  schema?: Schema<any, any>;
}
