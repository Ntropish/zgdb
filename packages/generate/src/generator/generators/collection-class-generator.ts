import { NormalizedSchema } from "../../parser/types.js";
import { IGenerator } from "./interface.js";

function generateSingleCollectionClass(schema: NormalizedSchema): string {
  if (schema.isJoinTable) {
    return "";
  }

  const collectionName = `${schema.name}Collection<TActor>`;
  const createInputName = `Create${schema.name}Input`;
  const nodeName = `${schema.name}Node<TActor>`;
  const fbsNodeName = `${schema.name}FB.${schema.name}`;
  const fbsRootGetter = `getRootAs${schema.name}`;

  const sortedFields = [...schema.fields].sort((a, b) =>
    a.name.localeCompare(b.name)
  );

  const createStrings = sortedFields
    .filter((f) => f.type === "string")
    .map(
      (f) =>
        `    const ${f.name}Offset = data.${f.name} ? builder.createString(data.${f.name}) : 0;`
    )
    .join("\n");

  const startVector = (fieldName: string, num: string) =>
    `${fbsNodeName}.start${
      fieldName.charAt(0).toUpperCase() + fieldName.slice(1)
    }Vector(builder, ${num});`;

  const createListVectors = sortedFields
    .filter((f) => f.isVector && f.type !== "string") // Assuming string arrays are handled differently or not at all for now
    .map((f) => {
      const fieldName = f.name;
      const count = `data.${fieldName}?.length ?? 0`;
      return `    ${startVector(fieldName, count)}
    for (const item of (data.${fieldName} ?? []).reverse()) {
      // numbers are added directly, other types might need builder offsets
      builder.add${f.type.charAt(0).toUpperCase() + f.type.slice(1)}(item);
    }
    const ${fieldName}VectorOffset = builder.endVector();`;
    })
    .join("\n\n");

  const addFields = sortedFields
    .map((f, index) => {
      const capName = f.name.charAt(0).toUpperCase() + f.name.slice(1);
      const param = f.isVector
        ? `${f.name}VectorOffset`
        : f.type === "string"
        ? `${f.name}Offset`
        : `data.${f.name}`;

      const defaultValue = f.type === "string" || f.isVector ? "0" : "null";
      return `${fbsNodeName}.add${capName}(builder, ${param} ?? ${defaultValue});`;
    })
    .join("\n    ");

  return `export class ${collectionName} {
    constructor(
      private db: ZgDatabase,
      private authContext: ZgAuthContext<TActor> | null
    ) {}

    get collectionName(): string {
      return '${schema.name}';
    }

    create(data: ${createInputName}): ${nodeName} {
      const builder = new Builder(1024);
      
      ${createStrings}
      ${createListVectors}

      ${fbsNodeName}.start${schema.name}(builder);
      
      ${addFields}

      const entityOffset = ${fbsNodeName}.end${schema.name}(builder);
      
      builder.finish(entityOffset);
      const buffer = builder.asUint8Array();

      this.db.insert(this.collectionName, data.id, buffer);

      const fbb = ${fbsNodeName}.${fbsRootGetter}(new ByteBuffer(buffer));

      return new ${nodeName}(
        this.db,
        fbb,
        this.authContext,
      );
    }

    *[Symbol.iterator](): Generator<${nodeName}> {
      for (const [id, buffer] of this.db.scan(this.collectionName)) {
        const fbb = ${fbsNodeName}.${fbsRootGetter}(new ByteBuffer(buffer));
        yield new ${nodeName}(
          this.db,
          fbb,
          this.authContext,
        );
      }
    }
  }`;
}

export class CollectionClassGenerator implements IGenerator {
  generate(schemas: NormalizedSchema[]): string {
    const classes = schemas
      .map((s) => generateSingleCollectionClass(s))
      .filter(Boolean)
      .join("\n\n");

    return `// --- Collection Classes ---\n${classes}`;
  }
}
