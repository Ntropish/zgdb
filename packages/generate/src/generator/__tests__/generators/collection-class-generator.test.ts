import { describe, it, expect } from "vitest";
import { CollectionClassGenerator } from "../../generators/collection-class-generator.js";
import { NormalizedSchema } from "../../../parser/types.js";

const testSchema: NormalizedSchema[] = [
  {
    name: "User",
    fields: [
      { name: "id", type: "string", required: true, attributes: new Map() },
      {
        name: "displayName",
        type: "string",
        required: true,
        attributes: new Map(),
      },
      {
        name: "createdAt",
        type: "long",
        required: true,
        attributes: new Map(),
      },
    ],
    indexes: [],
    relationships: [],
    manyToMany: [],
  },
];

describe("CollectionClassGenerator", () => {
  it("should generate correct collection classes", () => {
    const generator = new CollectionClassGenerator();
    const result = generator.generate(testSchema);
    expect(result).toMatchSnapshot();
  });
});
