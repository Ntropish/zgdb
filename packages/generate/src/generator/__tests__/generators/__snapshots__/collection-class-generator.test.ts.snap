// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CollectionClassGenerator > should generate correct collection classes 1`] = `
"// --- Collection Classes ---
export class UserCollection<TActor> {
    constructor(
      private db: ZgDatabase,
      private authContext: ZgAuthContext<TActor> | null
    ) {}

    get collectionName(): string {
      return 'User';
    }

    create(data: CreateUserInput): UserNode<TActor> {
      const builder = new Builder(1024);
      
          const displayNameOffset = data.displayName ? builder.createString(data.displayName) : 0;
    const idOffset = data.id ? builder.createString(data.id) : 0;
      

      UserFB.User.startUser(builder);
      
      UserFB.User.addCreatedAt(builder, data.createdAt ?? null);
    UserFB.User.addDisplayName(builder, displayNameOffset ?? 0);
    UserFB.User.addId(builder, idOffset ?? 0);

      const entityOffset = UserFB.User.endUser(builder);
      
      builder.finish(entityOffset);
      const buffer = builder.asUint8Array();

      this.db.insert(this.collectionName, data.id, buffer);

      const fbb = UserFB.User.getRootAsUser(new ByteBuffer(buffer));

      return new UserNode<TActor>(
        this.db,
        fbb,
        this.authContext,
      );
    }

    *[Symbol.iterator](): Generator<UserNode<TActor>> {
      for (const [id, buffer] of this.db.scan(this.collectionName)) {
        const fbb = UserFB.User.getRootAsUser(new ByteBuffer(buffer));
        yield new UserNode<TActor>(
          this.db,
          fbb,
          this.authContext,
        );
      }
    }
  }"
`;
